### PR TITLE
feat: load directories.bin as a bin object

### PR DESCRIPTION
**Pull request by isaacs**
Fix: https://github.com/npm/arborist/issues/296

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,15 +1,56 @@
...
+
+  const m = data.directories && data.directories.bin
+  if (!m || typeof m !== 'string')
+    return data
+
+  // cut off any monkey business, like setting directories.bin
+  // to ../../../etc/passwd or /etc/passwd or something like that.
+  const root = dirname(path)
+  const dir = join('.', join('/', m))
+  data.bin = await walkBinDir(root, dir, {})
```
Comment (nlf) [inline index.js:32]: so we if `directories.bin` points to a directory that contains directories, we flatten everything in all of those directories to a list of bin names? what if there's a collision? this feels like it should just be a readdir and not bother traversing subdirectories to me

**Comment by wraithgar** [comment]
This will also fix https://github.com/npm/cli/issues/3446

File: index.js (hunk 1 above)
Comment (isaacs) [reply index.js:32]: I agree.  But this is how it worked in v6 and before, so I thought it best to leave it as it is.  I'd suggest dropping support for this entirely in npm version 8 (or deprecating it in v8 and dropping in v9).

**Comment by robin-drexler** [comment]
@wraithgar could you explain how exactly this PR fixes npm/cli#3446? I'm just curious. 

If I understood this issue correctly it was caused by bins of optionalDependencies being deleted when another package that was not deleted had a bin with the same name. https://github.com/npm/cli/issues/3446#issuecomment-881808689

Thanks a lot!

**Comment by isaacs** [comment]
Ah, yeah, it seems that npm/cli#3446 is a false hit here.  That one isn't relying on `directories.bin`, it's something else.  I'll re-open.

**Pull request closed without merging**